File: src/cpp/asynchrony/05_run_heavy_functions.cpp
#include <chrono>
#include <future>
#include <iostream>
#include <random>
#include <string>
#include <thread>
#include <vector>
#include <atomic>

// Function to simulate a heavy computation
int blocking_function(int i) {
  try {
    std::cout << "Running blocking function " << i << std::endl;

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0, 1);

    // Simulate a heavy computation using sleep
    std::this_thread::sleep_for(std::chrono::seconds(10));

    std::cout << "Done running blocking function " << i << std::endl;
    return i;
  } catch (const std::exception &e) {
    std::cerr << "Exception in blocking function " << i << ": " << e.what() << '\n';
    return -1;
  }
}

// Function to handle user input
void user_input(std::atomic<bool> &running) {
  std::string user_text;
  while (running) {
    std::cout << "Enter some text: ";
    if (std::getline(std::cin, user_text)) {
      std::cout << "User input: " << user_text << std::endl;
    } else {
      running = false;
    }
  }
}

int main() {
  std::vector<std::future<int>> blocking_tasks;
  for (int i = 0; i < 6; ++i) {
    blocking_tasks.push_back(
        std::async(std::launch::async, blocking_function, i));
  }

  std::atomic<bool> running(true);
  std::thread input_thread(user_input, std::ref(running));

  std::vector<int> results;
  for (auto &task : blocking_tasks) {
    try {
      results.push_back(task.get());
    } catch (const std::exception &e) {
      std::cerr << "Exception in main: " << e.what() << '\n';
      results.push_back(-1);
    }
  }

  std::cout << "Results: ";
  for (const auto &result : results) {
    std::cout << result << " ";
  }
  std::cout << std::endl;

  running = false;
  input_thread.join();

  return 0;
}

File: src/python/asynchrony/05_run_heavy_functions.py
"""
This example demonstrates how to run CPU-bound tasks concurrently using asyncio
in combination with a ProcessPoolExecutor. This is useful when you want to achieve
concurrency with asyncio but have CPU-bound tasks that would otherwise block the
async event loop. The example shows how to offload these tasks to separate processes
managed by the ProcessPoolExecutor, allowing the event loop to continue processing
other async tasks without being blocked, such as receiving user input and immediately
displaying it.
"""

import asyncio
import concurrent.futures
import time
import numpy as np


def blocking_function(i):
    print(f"Running blocking function {i}")
    a = np.random.rand(3000, 3000)
    b = np.random.rand(3000, 3000)
    c = np.dot(a, b)
    print(f"Done running blocking function {i}")
    return i


async def user_input():
    while True:
        user_text = await asyncio.to_thread(input, "Enter some text: \n")
        print(f"User just input: {user_text}")


async def run_blocking_functions(executor, callback):
    print("Start run_blocking_functions")

    loop = asyncio.get_running_loop()
    blocking_tasks = [loop.run_in_executor(executor, callback, i) for i in range(10)]
    print("Waiting for executor tasks")
    completed, _ = await asyncio.wait(blocking_tasks)
    results = [task.result() for task in completed]
    print(f"Results: {results}")
    print("End run_blocking_functions")


async def main():
    executor = concurrent.futures.ProcessPoolExecutor(max_workers=3)

    await asyncio.gather(
        run_blocking_functions(executor, blocking_function),
        user_input(),
    )


if __name__ == "__main__":
    asyncio.run(main())
